packages/@sanity/base/src/datastores/grants/createGrantsStore.ts

```diff
@@ -0,0 +1,96 @@
+import {defer, of} from 'rxjs'
+
+import {mergeMap, publishReplay, switchMap} from 'rxjs/operators'
+import {evaluate, parse} from 'groq-js'
+import {SanityDocument} from '@sanity/types'
+import {refCountDelay} from 'rxjs-etc/operators'
+import {vxClient as sanityClient} from '../../client/versionedClient'
+import {
+  DatasetGrants,
+  DOCUMENT_FILTER_RULE_KEY,
+  GrantsStore,
+  DocumentPermissionName,
+  DocumentFilterRule,
+  PermissionCheckResult,
+} from './types'
+import {debugGrants$} from './debug'
+
+function fetchApiEndpoint<T>(endpoint: string): Promise<T> {
+  return sanityClient.request({
+    uri: endpoint,
+    withCredentials: true,
+  })
+}
+
+function getDatasetGrants(projectId: string, dataset: string): Promise<DatasetGrants> {
+  return fetchApiEndpoint(`/projects/${projectId}/datasets/${dataset}/grants`)
+}
+
+const PARSED_FILTERS_MEMO = new Map()
+async function matchesFilter(filter: string, document: SanityDocument) {
+  // console.time(`matchesFilter ${filter}`)
+  if (!PARSED_FILTERS_MEMO.has(filter)) {
+    PARSED_FILTERS_MEMO.set(filter, parse(`*[${filter}]`))
+  }
+  const parsed = PARSED_FILTERS_MEMO.get(filter)
+  const data = await (await evaluate(parsed, {dataset: [document]})).get()
+  // console.timeEnd(`matchesFilter ${filter}`)
+  return data?.length === 1
+}
+
+export function createGrantsStore(): GrantsStore {
+  const datasetGrants = defer(() => of(sanityClient.config())).pipe(
+    mergeMap(({projectId, dataset}) => {
+      if (!projectId || !dataset) {
+        throw new Error('Missing projectId or dataset')
+      }
+      return getDatasetGrants(projectId, dataset)
+    })
+  )
+  const currentUserDatasetGrants = debugGrants$.pipe(
+    switchMap((debugGrants) => (debugGrants ? of(debugGrants) : datasetGrants)),
+    publishReplay(1),
+    refCountDelay(1000)
```

Comment:
- `publishReplay(1)` will replay the latest value to new subscribers, similar to how `shareReplay({bufferCount: 1})` works
- `refCountDelay(1000)` is a variant of `refCount()` that waits for a the given amount of `ms` after subscription count drops to zero before unsubscribing from the source. This is useful when you have a situation where a new subscriber usually subscribes shortly after refcount has gone to zero. For example, when navigating between pages, usually all components gets unmounted and unsubscribe from their data sources. Shortly after, the different page components may start resubscribing to the same streams again, and `refCountDelay(1000)`, makes sure we keep the source subscribed to in between the unsubscribes and resubscribes. This is in particular useful when the source stream represents a long lived connection, (e.g. a listener request), since we then would avoid disconnecting it before reconnecting again the moment after.